File: src/editor/docks/node_tagging_dock.rs
use godot::{
    engine::{EditorInterface, IVBoxContainer, VBoxContainer},
    prelude::*,
};

use crate::{
    editor::{tag_dictionary_fs::TagDictionaryFs, ui::tag_tree::TagTree},
    tag_manager::TagManager,
};

#[derive(GodotClass)]
#[class(tool, init, base = VBoxContainer)]
pub struct NodeTaggingDock {
    base: Base<VBoxContainer>,
}

#[godot_api]
impl NodeTaggingDock {
    #[func]
    fn on_visibility_changed(&mut self) {
        if !self.to_gd().is_visible() {
            return;
        }

        // clear all children when the dock is visible
        for mut child in self.to_gd().get_children().iter_shared() {
            child.queue_free();
        }

        self.render_tag_trees();
    }

    #[func]
    fn on_tags_added(&mut self, tags: PackedStringArray) {
        let tag_manager = TagManager::new_alloc();

        if let Some(mut selection) = EditorInterface::singleton().get_selection() {
            for node in selection.get_selected_nodes().iter_shared() {
                tag_manager.bind().add_tags(tags.clone(), node);
            }
            EditorInterface::singleton().mark_scene_as_unsaved();
        }
    }

    #[func]
    fn on_tags_removed(&mut self, tags: PackedStringArray) {
        let tag_manager = TagManager::new_alloc();

        if let Some(mut selection) = EditorInterface::singleton().get_selection() {
            for node in selection.get_selected_nodes().iter_shared() {
                tag_manager.bind().remove_tags(tags.clone(), node);
            }
            EditorInterface::singleton().mark_scene_as_unsaved();
        }
    }

    #[func]
    fn render_tag_trees(&mut self) {
		// let's start with cleaning up child elements
		for mut child in self.to_gd().get_children().iter_shared() {
			child.queue_free();
		}

        let mut tag_dictionary_fs = TagDictionaryFs::new_gd();
        let tag_manager = TagManager::new_alloc();

        tag_dictionary_fs.bind_mut().scan_fs("res://".into());

        let dictionaries = tag_dictionary_fs.bind().get_dictionaries();

        for dict in dictionaries.iter_shared() {
            let mut tree = TagTree::new_alloc();

            if let Some(mut selection) = EditorInterface::singleton().get_selection() {
                for node in selection.get_selected_nodes().iter_shared() {
                    tree.bind_mut()
                        .set_selected_tags(tag_manager.bind().get_tags(node));
                }
            }

            tree.bind_mut().set_tag_dictionary(Some(dict.clone()));
            tree.bind_mut().set_selectable(true);
            tree.connect(
                "tags_added".into(),
                Callable::from_object_method(&self.to_gd(), "on_tags_added"),
            );
            tree.connect(
                "tags_removed".into(),
                Callable::from_object_method(&self.to_gd(), "on_tags_removed"),
            );

            self.to_gd().add_child(tree.to_variant().to());
        }
    }
}

#[godot_api]
impl IVBoxContainer for NodeTaggingDock {
    fn ready(&mut self) {
        self.to_gd().set_name("Node tags".into());

		if let Some(mut selection) = EditorInterface::singleton().get_selection() {
			selection.connect(
				"selection_changed".into(),
				Callable::from_object_method(&self.to_gd(), "render_tag_trees"),
			);
		}
		
    }
}

File: src/editor/docks/NodeTaggingDock.cpp
//
// Created by iamoc on 20/03/2025.
//

#include "NodeTaggingDock.h"

#include <godot_cpp/classes/editor_interface.hpp>
#include <godot_cpp/classes/editor_selection.hpp>

#include "../TagDictionaryFS.h"
#include "../../TagManager.h"


using namespace octod::gameplay::tags::editor::docks;

void NodeTaggingDock::_ready()
{
	VBoxContainer::_ready();
}

void NodeTaggingDock::_bind_methods()
{
}

void NodeTaggingDock::on_tags_added(const PackedStringArray &p_tags_added)
{
	const TagManager *tag_manager = TagManager::get_singleton();

	if (EditorSelection *editor_selection = EditorInterface::get_singleton()->get_selection())
	{
		const TypedArray<Node> selected_nodes = editor_selection->get_selected_nodes();

		for (int i = 0; i < selected_nodes.size(); i++)
		{
			tag_manager->add_tags(cast_to<Node>(selected_nodes[i]), p_tags_added);
		}
	}
}

void NodeTaggingDock::on_tags_removed(const PackedStringArray &p_tags_removed)
{
	const TagManager *tag_manager = TagManager::get_singleton();

	if (EditorSelection *editor_selection = EditorInterface::get_singleton()->get_selection())
	{
		const TypedArray<Node> selected_nodes = editor_selection->get_selected_nodes();

		for (int i = 0; i < selected_nodes.size(); i++)
		{
			tag_manager->remove_tags(cast_to<Node>(selected_nodes[i]), p_tags_removed);
		}
	}
}

void NodeTaggingDock::on_visibility_changed() const
{
	if (!is_visible())
	{
		return;
	}

	for (int i = 0; i < get_child_count(); i++)
	{
		const Ref child = get_child(i);
		child->queue_free();
	}
}

void NodeTaggingDock::render_tag_trees()
{
	for (int i = 0; i < get_child_count(); i++)
	{
		cast_to<Node>(get_child(i))->queue_free();
	}

	TagDictionaryFS::get_singleton()->scan_file_system("res://");

	const auto dictionaries = TagDictionaryFS::get_singleton()->get_dictionaries();

	for (int i = 0; i < dictionaries.size(); i++)
	{
		/// todo: port tag tree node
	}
}
